Fixed #24638 -- Added support for SQL comments

viciu opened the pull request:
This PR reactivates work done in #4495 and #15711 -  useful addition to Django for adding traceability for queries.

# Trac ticket number

ticket-24638

# Branch description

Add queryset comment() method

- [x] Rebase against 5.x
- [x] Support SELECT statement
- [x] Support UPDATE statement
- [ ] Support INSERT statement
- [ ] Support DELETE statement
- [x] use regex instead if "*/"


# Checklist

- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

hannseman committed:
Fixed #24638 -- Added support for SQL comments in SELECT and UPDATE queries.

Co-authored-by: Adam Johnson <me@adamj.eu>
Co-authored-by: Wiktor Kolodziej <wiktor@pykonik.org>

adamchainz commented on django/db/models/sql/compiler.py:
```diff
@@ -795,6 +795,9 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
                     result += distinct_result
                     params += distinct_params
 
+                if self.query.comments:
```
Let's put the coment right after SELECT, before the DISTINCT clause which can be arbitrarily long.

viciu replied:
Done, added a test as well.

adamchainz commented on django/db/models/sql/compiler.py:
```diff
@@ -1627,6 +1630,9 @@ def explain_query(self):
                     yield value
 
+    def comments_sql(self):
+        return " ".join([f"/* {comment} */" for comment in self.query.comments])
```
Can we avoid the join? Callers should all be doing it already.

viciu replied:
DONE

adamchainz reviewed (changes requested):
Great start, looking forward to your iteration tomorrow.

The PR description formatting needs a little fixing.

adamchainz commented on django/db/models/sql/compiler.py:
```diff
@@ -1627,6 +1630,9 @@ def explain_query(self):
                     yield value
 
+    def comments_sql(self):
+        return [f"/* {comment} */" for comment in self.query.comments]
```
another idea: perhaps we can add the `/* ... */` wrapping to comments in `QuerySet.comment()`. Then we wouldn't even need a `comments_sql` method, the tuple can just be thrown into the SQL list.

viciu replied:
Done

adamchainz commented on docs/ref/models/querysets.txt:
```diff
@@ -2030,6 +2030,37 @@ See the :doc:`/topics/db/sql` for more information.
+.. method:: comment(message)
+
+Adds an SQL comment to be inserted into the resultant SQL statement, after
+``SELECT``/``UPDATE``, and if applicable, ``DISTINCT``.. For example::
```
Is it not possible to make it work for `INSERT`, at least when using `Model.objects.comment(...).create(...)` ?

adamchainz commented on docs/ref/models/querysets.txt:
```diff
@@ -2030,6 +2030,37 @@ See the :doc:`/topics/db/sql` for more information.
+
+    SELECT /* blog/views.py:50 */ blog_entry.* FROM blog_entry
+
+This is useful for adding information to the SQL statement which when logged
+can be used to correlate slow and problematic queries to the generating source
+code.
```
```suggestion
This is useful for adding logging information to the generated SQL for
correlating problematic queries with the responsible Python code.
```

adamchainz commented on docs/ref/models/querysets.txt:
```diff
@@ -2030,6 +2030,37 @@ See the :doc:`/topics/db/sql` for more information.
+
+… translates into::
+
+    UPDATE blog_entry.* /* blog/views.py:75 */ /* python=3.11 */ SET "headline" = 'This is new'
```
Needs updating
```suggestion
    UPDATE /* blog/views.py:75 */ /* python=3.11 */ blog_entry SET "headline" = 'This is new'
```

adamchainz commented on docs/releases/5.2.txt:
```diff
@@ -178,7 +178,9 @@ Migrations
 ~~~~~~
 
-* ...
+* The new :meth:`.QuerySet.comment` method allows for injecting comments into
+  the generated SQL statements. This can be useful for correlating troublesome
+  queries to the application code.
```
Try to avoid judgmental words like “troublesome”, in favour of specific terms like “slow”. It’s more precise and translates better.

```suggestion
* The new :meth:`.QuerySet.comment` method allows for adding comments into
  the generated SQL. This can be useful for correlating slow queries with
  application code.
```

viciu committed:
Fixed #24638 -- Added support for SQL comments

- Moved wrapping to comments in QuerySet.comment()
- Removed comments_sql()
- Added re.search instead of use if */
- WIP INSERT and DELETE

adamchainz commented on django/db/models/query.py:
```diff
@@ -1783,6 +1787,18 @@ def using(self, alias):
+                "Cannot pass strings containing /* or */ to comment(). "
+                "Escape or strip these delimiters before calling comment()."
+            )
+        commented_message = f"/* {message} */".format(message=message)
```
It's an f-string, the `.format` isn't doing anything
```suggestion
        commented_message = f"/* {message} */"
```

adamchainz commented on tests/queries/tests.py:
```diff
@@ -4602,3 +4602,74 @@ def test_ticket_23622(self):
+        self.assertIn("UPDATE /* this is an update */ ", captured_queries[0]["sql"])
+
+    def test_stops_delimiters_in_message(self):
+        for comment in ["foo=/a/b/c*/", "/*foo=/a/b/c", "**//SELECT nothing;//**"]:
+            msg = (
+                "Cannot pass strings containing /* or */ to comment(). "
+                "Escape or strip these delimiters before calling comment()."
+            )
+
+            with self.subTest(comment), self.assertRaisesMessage(ValueError, msg):
+                NamedCategory.objects.comment(comment)
```
We can hoist `msg` and split the subtests one per line, to make it easier to read and add any more in the future.

```suggestion
        msg = (
            "Cannot pass strings containing /* or */ to comment(). "
            "Escape or strip these delimiters before calling comment()."
        )
        for comment in [
            "foo=/a/b/c*/",
            "/*foo=/a/b/c",
            "**//SELECT nothing;//**",
        ]:
            with self.subTest(comment), self.assertRaisesMessage(ValueError, msg):
                NamedCategory.objects.comment(comment)
```

adamchainz commented on tests/queries/tests.py:
```diff
@@ -4602,3 +4602,74 @@ def test_ticket_23622(self):
+
+    def test_delete(self):
+        with CaptureQueriesContext(connection) as captured_queries:
+            NamedCategory.objects.create(name="test")
+            NamedCategory.objects.comment("DELETE 1").all().delete()
+
+            NamedCategory.objects.create(name="test")
+            NamedCategory.objects.comment("DELETE 2").all().delete()
+        self.assertIn("DELETE FROM /* DELETE 1 */", captured_queries[1]["sql"])
```
1. We don't need to create any data in order to run `DELETE`
2. Only one delete query needed
3. Use `startswith` and more realistic example data

```suggestion
            NamedCategory.objects.comment("blog/views.py:50").all().delete()
        self.assertTrue(
            captured_queries[0]["sql"].startswith(
                "DELETE /* blog/views.py:50 */ FROM", 
            )
        )
```

viciu replied:
The problem here is that delete does not generate delete query, but select:

This

```
NamedCategory.objects.comment("DELETE 2").delete()
```

will fail:
```
FAIL: test_delete (queries.tests.SqlCommentsInQueriesTests.test_delete)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/viciu/prog/django/tests/queries/tests.py", line 4670, in test_delete
    self.assertIn("DELETE FROM /* DELETE 1 */", [q['sql'] for q in captured_queries])
AssertionError: 'DELETE FROM /* DELETE 1 */' not found in 
['SELECT /* DELETE 2 */ "queries_dumbcategory"."id", "queries_namedcategory"."dumbcategory_ptr_id", "queries_namedcategory"."name" FROM "queries_namedcategory" INNER JOIN "queries_dumbcategory" ON ("queries_namedcategory"."dumbcategory_ptr_id" = "queries_dumbcategory"."id")']

```


therefore I was experimenting with adding some records first and then deleting them. If I understand correctly it looks like django fetches records to delete first, then updates related records with nulls, then finally deletes them one by one (including related records):

```
FAIL: test_delete (queries.tests.SqlCommentsInQueriesTests.test_delete)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/viciu/prog/django/tests/queries/tests.py", line 4670, in test_delete
    self.assertIn("DELETE FROM /* DELETE 1 */", [q['sql'] for q in captured_queries])
AssertionError: 'DELETE FROM /* DELETE 1 */' not found in 
[
'INSERT INTO "queries_dumbcategory" ("id") VALUES (NULL)', 
'INSERT INTO "queries_namedcategory" ("dumbcategory_ptr_id", "name") VALUES (1, \'test\')', 
'SELECT /* DELETE 1 */ "queries_dumbcategory"."id", "queries_namedcategory"."dumbcategory_ptr_id", "queries_namedcategory"."name" FROM "queries_namedcategory" INNER JOIN "queries_dumbcategory" ON ("queries_namedcategory"."dumbcategory_ptr_id" = "queries_dumbcategory"."id")', 
'UPDATE "queries_tag" SET "category_id" = NULL WHERE "queries_tag"."category_id" IN (1)', 
'DELETE FROM "queries_namedcategory" WHERE "queries_namedcategory"."dumbcategory_ptr_id" IN (1)',
'DELETE FROM "queries_dumbcategory" WHERE "queries_dumbcategory"."id" IN (1)', 

'INSERT INTO "queries_dumbcategory" ("id") VALUES (NULL)', 
'INSERT INTO "queries_namedcategory" ("dumbcategory_ptr_id", "name") VALUES (2, \'test\')', 
'SELECT /* DELETE 2 */ "queries_dumbcategory"."id", "queries_namedcategory"."dumbcategory_ptr_id", "queries_namedcategory"."name" FROM "queries_namedcategory" INNER JOIN "queries_dumbcategory" ON ("queries_namedcategory"."dumbcategory_ptr_id" = "queries_dumbcategory"."id")', 
'UPDATE "queries_tag" SET "category_id" = NULL WHERE "queries_tag"."category_id" IN (2)', 
'DELETE FROM "queries_namedcategory" WHERE "queries_namedcategory"."dumbcategory_ptr_id" IN (2)', 
'DELETE FROM "queries_dumbcategory" WHERE "queries_dumbcategory"."id" IN (2)']
```

adamchainz commented:
Thanks for putting sprint time in on this! I found a bit more time here to review, hope you'll be able to update soon.

charettes commented on django/db/models/sql/compiler.py:
```diff
@@ -1891,7 +1895,13 @@ def contains_self_reference_subquery(self):
-        delete = "DELETE FROM %s" % self.quote_name_unless_alias(query.base_table)
+        result = [
+            "DELETE",
+            *self.query.comments,
```
What's the rationale for injecting comments here instead of at the end of the query?

viciu replied:
Thanks for your comment. Queries can be long and truncated - please refer to the comment https://github.com/django/django/pull/15711#issuecomment-1132633387

adamchainz replied:
Yeah, the idea is that in a running query list, you can see the type of query first, then any tracing info. I saw this recommended by MySQL experts years ago, hence my suggestion for it in the previous PR. This is not for supporting any MySQL specific options.

I'm not so sure now that we should try to support inserting comments at any of the positions used for query hints. That would require a very complex API and might still not cover all cases. Also, database instrumentation exists to allow rewriting queries when needed.

charettes replied:
> I saw this recommended by MySQL experts years ago.

Do you have references for that? Isn't it fair to assume that an approach that works well for MySQL and recommended by these experts might not work so well for other backends (e.g. Postgres). I've seen plenty of [solutions](https://github.com/google/sqlcommenter/blob/c447d218a1b8ff628571b22cf4e74f2a6fe2e38a/python/sqlcommenter-python/google/cloud/sqlcommenter/__init__.py#L46-L53) and APMs default to prepending the traces to queries to ensure they are never truncated.

At work we built a solution that postpends commented traces specifically for MySQL and it appears to be working well hence my curiosity here.

charettes replied:
From an internal discussion with DBREs

1. Some tooling doesn’t support parsing in the middle of a query
2. Keeping queries starting with their respective verb makes it easier to filter
3. Putting the trace at the end of query might cause it to be truncated off on some systems

This makes me think that a proper solution would at the very least need to be configurable?

charettes commented on django/db/models/sql/compiler.py:
```diff
@@ -1975,9 +1985,10 @@ def as_sql(self):
         result = [
-            "UPDATE %s SET" % qn(table),
+            "UPDATE",
+            *self.query.comments,
```
Ditto.

charettes reviewed (commented):
I understand the desire to help workaround MySQL'ism but I find it strange that we default to injecting comments where it can be useful for the latter on all backends given the main purpose of these is likely to be traceability which is harder to parse when it's right in the middle SQL.

Could we have some backends level option that at least configure where they should live and default to prepending / postpending them on non-MySQL/Oracle backends?

pgramaca commented:
Let me know if there is something i can help with.
I would love to help with this feature.

jacobtylerwalls commented:
Hi @viciu, we just got a new proposal in #21226. Were you planning to continue here, or should we continue with the new one?

Happy to keep iterating here if you like 👍
